main.go

```diff
@@ -604,35 +607,55 @@ func (c controller) waitForPod(ctx context.Context, name string) error {
 	})
 }
 
-//nolint:nestif
-func (c *controller) populate(hashrings []receive.HashringConfig, statefulsets map[string]*appsv1.StatefulSet) {
+func (c *controller) populate(ctx context.Context, hashrings []receive.HashringConfig, statefulsets map[string]*appsv1.StatefulSet) {
 	for i, h := range hashrings {
-		if sts, exists := statefulsets[h.Hashring]; exists {
-			var endpoints []string
+		sts, exists := statefulsets[h.Hashring]
+		if !exists {
+			continue
+		}
 
-			for i := 0; i < int(*sts.Spec.Replicas); i++ {
-				// If cluster domain is empty string we don't want dot after svc.
-				clusterDomain := ""
-				if c.options.clusterDomain != "" {
-					clusterDomain = fmt.Sprintf(".%s", c.options.clusterDomain)
+		var endpoints []string
+
+		for i := 0; i < int(*sts.Spec.Replicas); i++ {
+			if c.options.allowDynamicScaling {
+				podName := fmt.Sprintf("%s-%d", sts.Name, i)
+
+				pod, err := c.klient.CoreV1().Pods(c.options.namespace).Get(ctx, podName, metav1.GetOptions{})
+				if kerrors.IsNotFound(err) {
+					continue
+				}
+				// Do not add a replica to the hashring if pod is not Ready.
+				if !podutils.IsPodReady(pod) {
+					level.Warn(c.logger).Log("msg", "failed adding pod to hashring, pod not ready", "pod", podName, "err", err)
+					continue
 				}
 
-				endpoints = append(endpoints,
-					fmt.Sprintf("%s-%d.%s.%s.svc%s:%d",
-						sts.Name,
-						i,
-						sts.Spec.ServiceName,
-						c.options.namespace,
-						clusterDomain,
-						c.options.port,
-					),
-				)
+				if pod.ObjectMeta.DeletionTimestamp != nil && (pod.Status.Phase == corev1.PodRunning || pod.Status.Phase == corev1.PodPending) {
+					// Pod is terminating, do not add it to the hashring.
+					continue
+				}
+			}
```

Comment:
Maybe we can reuse `waitForPod` here? But add the `pod.ObjectMeta.DeletionTimestamp != nil` check to it?

Reply:
Hm do we want to poll the pod though?